### PR TITLE
[JENKINS-48027] Properly encode branch names

### PR DESCRIPTION
**Pull request by imeredith**
# Description

See [JENKINS-48027](https://issues.jenkins-ci.org/browse/JENKINS-48027).

# Submitter checklist
- [ ] Link to JIRA ticket in description, if appropriate.
- [ ] Change is code complete and matches issue description
- [ ] Appropriate unit or acceptance tests or explanation to why this change has no tests
- [ ] Reviewer's manual test instructions provided in PR description. See Reviewer's first task below.

# Reviewer checklist
- [ ] Run the changes and verified the change matches the issue description
- [ ] Reviewed the code
- [ ] Verified that the appropriate tests have been written or valid explanation given

### CODE REVIEW

File: blueocean-core-js/src/js/paths/rest.js [hunk 1]
```diff
@@ -28,7 +29,7 @@ export default {
 
     run({ organization, pipeline, branch, runId }) {
         if (branch) {
-            return `${this.pipeline(organization, pipeline)}branches/${encodeURIComponent(encodeURIComponent(branch))}/runs/${runId}/`;
+            return `${this.pipeline(organization, pipeline)}branches/${encodeURIComponent(JenkinsEncode.encode(branch))}/runs/${runId}/`;
```
Comment (kzantow) [inline blueocean-core-js/src/js/paths/rest.js]: ughhhhhhhhhhhh this is exactly what I was talking about: just building strings for the urls

File: blueocean-core-js/src/js/services/ActivityService.js [hunk 2]
```diff
@@ -72,6 +72,7 @@ export class ActivityService extends BunkerService {
                 // Should really have dedupe on methods like these, but for now
                 // just clone data so that we dont modify other instances.
                 const run = utils.clone(data);
+                console.log('ba', run);
```
Comment (kzantow) [inline blueocean-core-js/src/js/services/ActivityService.js]: 🐛 no console

File: blueocean-core-js/src/js/utils/jenkins-encode.js [hunk 3]
```diff
@@ -0,0 +1,31 @@
+export default class JenkinsEncode {
+    static _uriMap = null;
+    static _getUrlMap() {
+        if(this._uriMap == null) {
+            const raw = "!  $ &'()*+,-. 0123456789   =  @ABCDEFGHIJKLMNOPQRSTUVWXYZ    _ abcdefghijklmnopqrstuvwxyz";
```
Comment (kzantow) [inline blueocean-core-js/src/js/utils/jenkins-encode.js]: :bug: this stuff should just be a static property somewhere

Comment (kzantow) [reply blueocean-core-js/src/js/utils/jenkins-encode.js]: why so many spaces?

File: blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/BranchImpl.java [hunk 4]
```diff
@@ -62,7 +64,11 @@ public Branch getBranch() {
 
     @Override
     public Link getLink() {
-        return parent.rel(Util.rawEncode(getName()));
+        try {
+            return parent.rel(URLEncoder.encode(getName(), "UTF-8").replace("+", "%20"));
```
Comment (kzantow) [inline blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/BranchImpl.java]: I still don't understand this part

File: blueocean-core-js/src/js/utils/jenkins-encode.js [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+              this._uriMap[i] = (raw.charAt(j) == ' ');
+            }
+        }
+        return this._uriMap;
+    }
+
+    static encode(str) {
+        const uriMap = this._getUrlMap();
+
+        return str.split('').map(c => {
```
Comment (kzantow) [inline blueocean-core-js/src/js/utils/jenkins-encode.js]: this seems really inefficient

File: blueocean-core-js/src/js/utils/jenkins-encode.js (hunk 3 above)
Comment (imeredith) [reply blueocean-core-js/src/js/utils/jenkins-encode.js]: Spaces are the character placeholders where we want to encode lol

File: blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/BranchImpl.java (hunk 4 above)
Comment (imeredith) [reply blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/BranchImpl.java]: What part is that?

File: blueocean-core-js/src/js/utils/jenkins-encode.js (hunk 5 above)
Comment (kzantow) [reply blueocean-core-js/src/js/utils/jenkins-encode.js]: but only a few characters, so fine

File: blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/BranchImpl.java (hunk 4 above)
Comment (kzantow) [reply blueocean-pipeline-api-impl/src/main/java/io/jenkins/blueocean/rest/impl/pipeline/BranchImpl.java]: ahhh, got it, incomplete encoding

File: blueocean-core-js/src/js/utils/jenkins-encode.js (hunk 3 above)
Comment (kzantow) [reply blueocean-core-js/src/js/utils/jenkins-encode.js]: oooh. funny

**Comment by kshultzCB** [comment]
LGTM too. :bee:

**Pull request merged**